SetupDataPkg/Library/ConfigVariableListLib/ConfigVariableListLib.c

```diff
@@ -0,0 +1,292 @@
+    CRC32      = *(UINT32 *)(DataInBin + VarList->DataSize);
+
+    // on next iteration, skip past this variable
+    ListIndex += sizeof (*VarList) + VarList->NameSize + VarList->DataSize + sizeof (*Guid) + sizeof (Attributes)
+                 + sizeof (CRC32);
+
+    VarName = AllocatePool (VarList->NameSize);
+    if (VarName == NULL) {
+      DEBUG ((DEBUG_ERROR, "%a Failed to allocate memory for VarName size: %u\n", __FUNCTION__, VarList->NameSize));
+      Status = EFI_OUT_OF_RESOURCES;
+      goto Exit;
+    }
+
+    CopyMem (VarName, NameInBin, VarList->NameSize);
+
+    // This check needs to come after we allocate VarName so we can compare against a 16 bit aligned
+    // string and not assert
+    if (ConfigVarName && (0 != StrnCmp (ConfigVarName, VarName, VarList->NameSize / 2))) {
+      // Not the entry we are looking for
+      // While we are here, set the Name entry to NULL so we can tell if we found the entry later
+      (*ConfigVarListPtr)->Name = NULL;
+      FreePool (VarName);
+      continue;
+    }
+
+    Data = AllocatePool (VarList->DataSize);
+    if (Data == NULL) {
+      DEBUG ((DEBUG_ERROR, "%a Failed to allocate memory for Data size: %u\n", __FUNCTION__, VarList->DataSize));
+      Status = EFI_OUT_OF_RESOURCES;
+      // Free VarName here, as other memory gets freed in exit routine, but we do not increment
+      // *ConfigVarListCount for this entry
+      FreePool (VarName);
+      goto Exit;
+    }
+
+    CopyMem (Data, DataInBin, VarList->DataSize);
+
+    // Add correct values to this entry in the blob
+    Entry = &(*ConfigVarListPtr)[*ConfigVarListCount];
+
+    Entry->Name       = VarName;
+    Entry->Guid       = *Guid;
+    Entry->Attributes = Attributes;
+    Entry->Data       = Data;
+    Entry->DataSize   = VarList->DataSize;
+    (*ConfigVarListCount)++;
+
+    if (ConfigVarName) {
+      // Found the entry we are looking for
+      break;
+    }
+  }
+
+  if (!ConfigVarName) {
+    // No need to realloc for single entry
+    *ConfigVarListPtr = ReallocatePool (BinSize, sizeof (CONFIG_VAR_LIST_ENTRY) * *ConfigVarListCount, *ConfigVarListPtr);
+    if (!*ConfigVarListPtr) {
+      DEBUG ((DEBUG_ERROR, "%a Failed to reallocate memory for ConfigVarListPtr size: %u\n", __FUNCTION__, BinSize));
+      return EFI_OUT_OF_RESOURCES;
+    }
+  } else if (!(*ConfigVarListPtr)->Name) {
+    // We did not find the entry in the var list
+    // caller is responsible for freeing input
+    DEBUG ((DEBUG_ERROR, "%a Failed to find varname in var list: %s\n", __FUNCTION__, ConfigVarName));
+    return EFI_NOT_FOUND;
+  }
+
+Exit:
+  if (EFI_ERROR (Status)) {
+    // Need to free all allocated memory
+    while (*ConfigVarListCount >= 0) {
+      FreePool ((*ConfigVarListPtr)[*ConfigVarListCount - 1].Name);
+      FreePool ((*ConfigVarListPtr)[*ConfigVarListCount - 1].Data);
+      *ConfigVarListCount--;
+    }
+
+    FreePool (*ConfigVarListPtr);

```

Comment:
I think the loop above will reset the counter, but should we set the pointer to NULL as well?

Reply:
For FreePool functions, it is always good to check that it is not already NULL before freeing.